source/common/event/libevent_scheduler.cc

```diff
@@ -41,5 +49,54 @@ void LibeventScheduler::run(Dispatcher::RunType mode) {
 
 void LibeventScheduler::loopExit() { event_base_loopexit(libevent_.get(), nullptr); }
 
+void LibeventScheduler::initializeStats(DispatcherStats* stats) {
+  stats_ = stats;
+  evwatch_prepare_new(libevent_.get(), &onPrepare, this);
+  evwatch_check_new(libevent_.get(), &onCheck, this);
+}
+
+void LibeventScheduler::onPrepare(evwatch*, const evwatch_prepare_cb_info* info, void* arg) {
+  // `self` is `this`, passed in from evwatch_prepare_new.
+  auto self = static_cast<LibeventScheduler*>(arg);
+
+  // Record poll timeout and prepare time for this iteration of the event loop. The timeout is the
+  // expected polling duration, whereas the actual polling duration will be the difference measured
+  // between the prepare time and the check time immediately after polling. These are compared in
+  // onCheck to compute the poll_delay stat.
+  self->timeout_set_ = evwatch_prepare_get_timeout(info, &self->timeout_);
+  evutil_gettimeofday(&self->prepare_time_, NULL);
+
+  // If we have a check time available from a previous iteration of the event loop (that is, all but
+  // the first), compute the loop_duration stat.
+  if (self->check_time_.tv_sec != 0) {
+    timeval delta;
+    evutil_timersub(&self->prepare_time_, &self->check_time_, &delta);
+    recordTimeval(self->stats_->loop_duration_us_, delta);
+  }
+}
+
+void LibeventScheduler::onCheck(evwatch*, const evwatch_check_cb_info*, void* arg) {
+  // `self` is `this`, passed in from evwatch_check_new.
+  auto self = static_cast<LibeventScheduler*>(arg);
+
+  // Record check time for this iteration of the event loop. Use this together with prepare time
+  // from above to compute the actual polling duration, and store it for the next iteration of the
+  // event loop to compute the loop duration.
+  evutil_gettimeofday(&self->check_time_, NULL);
+  if (self->timeout_set_) {
+    timeval delta, delay;
+    evutil_timersub(&self->check_time_, &self->prepare_time_, &delta);
+    evutil_timersub(&delta, &self->timeout_, &delay);
+
+    // Delay can be negative, meaning polling completed early. This happens in normal operation,
+    // either because I/O was ready before we hit the timeout, or just because the kernel was
+    // feeling saucy. Disregard negative delays in stats, since they don't indicate anything
```

Comment:
Should we discard or record a 0 delay when negative? I'm thinking the latter could be good to ensure we don't get anomalies like the following. Imagine there is a slight negative delay normally, and you see this on a million iterations. But, scattered amongst these, we see 10 iterations where there is an elevated delay. These are really only a tiny fraction and don't indicate a huge problem, but they will appear as 100% with elevated delay in the histogram if we discard here.

Reply:
Recording the early wakeups as 0 delay would give you a number that means something different from what we're looking for. It might be worth recording as well (although I'm not sure what name I'd give it), but the thing we care about most here is specifically the distribution of positive delays.

One thing I tried to be very clear about in the docs is that delays are normal; they don't indicate a problem if they're within some expected range. For example, if you see the `poll_delay_us` number floating around a millisecond (or whatever your time slice is), with no long tail, then you're happy. It doesn't mean that every iteration of the event loop has a millisecond delay, it just means that when there is a delay, it's reasonably bounded.

If we included early delays as zeroes, it would skew the histogram such that 0 appears to be the correct ideal delay, and that the 1ms delay appears to be problematic tail latency.

Reply:
What I'm struggling with here is the difference between -1ms, 0ms and 1ms delays. To me, they seem basically the same thing, the kernel woke you up roughly when you expected. Or, is there something special about -1ms vs 0ms?

Reply:
Ah, yeah. The main difference is that libevent treats a just-slightly-too-early wakeup the same as a way-too-early wakeup: it doesn't process the timeout event and just goes back to sleep. It sometimes does this a few iterations in a row, sleeping for 0us on each, before finally sleeping for a bit longer.

One way to avoid the difference is to have a configurable minimum threshold. For example, don't record any delays less than 100us, because they're not interesting. But I don't know if I'd actually recommend implementing that, at least not until we've observed how this behaves in production.